File: auto_claim.php
<?php

/*
This script claims any claimable balance of a trusted asset.
Assets without trustline will be ignored.

In this model we write a python script for each wallet listed in the database.
If there is no private key, unsigned XDR will be print on the screen.
*/

$contas=json_decode(file_get_contents("wallets_and_keys.json"),true);

foreach($contas as $pub => $pvt) {
   verificarConta($pub,$pvt);
}

function verificarConta($myAddr,$pvt) {
   echo "Checking wallet $myAddr\n";
   $claimable = get_page("https://horizon.stellar.org/claimable_balances/?limit=200&claimant=$myAddr");
   $claimable=$claimable["_embedded"]["records"];
   $wlt="[" . substr($myAddr,-10) . "]";
   if (count($claimable) > 0) {
      $trustlines=array();
      $ids_claims=array();
      $account=get_page("https://horizon.stellar.org/accounts/$myAddr");
      $balances=$account["balances"];
      for ($i=0;$i<count($claimable);$i++) {
         $asset=explode(":",$claimable[$i]["asset"]);
         echo "\n$wlt (" . $asset[0] . ") ";
         //Verificar se possui trustline pro asset
         $possui_trustline=false;
         $claimants=$claimable[$i]["claimants"];
         $pode_claim=check_claimants($claimants,$myAddr);
         for ($j=0;$j<count($balances);$j++){
            if (($balances[$j]["asset_code"] == $asset[0]) && ($balances[$j]["asset_issuer"] == $asset[1])){
               echo "☑️"; // Já possui trustline
               if ($pode_claim) {
                  echo $claimable[$i]["id"];
                  $ids_claims[]=$claimable[$i]["id"];
               }
               $possui_trustline=true;
            }
         }
/*         if (!$possui_trustline && $pode_claim) {
            if (strlen($asset[0]) <= 4) {
               $ass_type="credit_alphanum4";
            }
            else {
               $ass_type="credit_alphanum12";
            }
            $cotacao=get_page("https://horizon.stellar.org/order_book?selling_asset_type=$ass_type&selling_asset_code=" . $asset[0] . "&selling_asset_issuer=" . $asset[1] . "&buying_asset_type=native");
            $cotacao=$cotacao["bids"];
            if (count($cotacao) > 0) {
               $preco = $cotacao[0]["price"];
               $valor_mercado=$claimable[$i]["amount"]*$preco;
               echo "Preco: $preco XLM x " . $claimable[$i]["amount"] . " = $valor_mercado XLM\n";
               echo $claimable[$i]["asset"] . " - " . $claimable[$i]["id"];
               $id_asset=str_replace(":","_",$claimable[$i]["asset"]);
               $trustlines[$id_asset]=$claimable[$i]["asset"];
               $ids_claims[]=$claimable[$i]["id"];
            }
            else {
               echo $asset[0] . " sem ofertas.";
            }
         }*/
      }
   }
   else {
      echo "0️⃣ no claimable.";
   }
   echo "\n";
   if (count($ids_claims) > 0) {
      proceed_trans($myAddr,array(),$ids_claims,$pvt);
   }
}

function check_claimants($claimants,$myAddr){
   $pode_claim=false;
   $now = new DateTime();
   $agora=time();
   for ($k=0;$k<count($claimants);$k++) {
      if ($claimants[$k]["destination"] == $myAddr) {
         if ($claimants[$k]["predicate"]["unconditional"] == 1) {
            echo "‼️ unconditional.";
            $pode_claim=true;
         }
         elseif (isset($claimants[$k]["predicate"]["and"])) {
            $ate_quando=$claimants[$k]["predicate"]["and"][1]["abs_before_epoch"];
            $apartir_quando=$claimants[$k]["predicate"]["and"][0]["not"]["abs_before_epoch"];
            if ($apartir_quando > $agora) {
               $pode_claim=false;
               echo "⏳ wait " . secondsToTime($apartir_quando);
            }
            elseif($ate_quando < $agora) {
               $pode_claim=false;
               echo "💸 expired";
            }
            else {
               echo "✅ ($apartir_quando < $agora < $ate_quando) available ";
               $pode_claim=true;
            }
         }
         elseif (isset($claimants[$k]["predicate"]["abs_before_epoch"])) {
            $ate_quando=$claimants[$k]["predicate"]["abs_before_epoch"];
            if($ate_quando < $agora) {
               $pode_claim=false;
               echo "❌ expired";
            }
            else {
               echo "✅ available.";
               $pode_claim=true;
            }
         }
         elseif (isset($claimants[$k]["predicate"]["not"]["abs_before_epoch"])) {
            $apartir_quando=$claimants[$k]["predicate"]["not"]["abs_before_epoch"];
            if ($apartir_quando > $agora) {
               $pode_claim=false;
               echo "⏳ wait " . secondsToTime($apartir_quando);
            }
            else {
               echo "✅ available";
               $pode_claim=true;
            }
         }
      }
   }
   return $pode_claim;
}

function get_page($url) {
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, True);
    curl_setopt($curl, CURLOPT_URL, $url);
    curl_setopt($curl,CURLOPT_USERAGENT,'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:7.0.1) Gecko/20100101 Firefox/7.0.1');
    $return = curl_exec($curl);
    curl_close($curl);
    return json_decode($return,true);
}

function proceed_trans($account,$trusts,$claims,$myPrivateKey=null) {
   //Path for the temporary pythin generated file below:
   $caminho="/Users/renato/Downloads/exec_" . uniqid() . ".py";

   $gravar=fopen($caminho,"w+");
   fwrite($gravar,'from stellar_sdk import Server, Keypair, TransactionBuilder, Network, FeeBumpTransaction, ClaimPredicate, Claimant, Asset
import requests

# 1. Load Keys
server = Server("https://horizon.stellar.org")
print("Gerando XDR...")

base_fee = 111
account = server.load_account("' . $account . '")

transaction = TransactionBuilder(
   source_account=account,
   network_passphrase=Network.PUBLIC_NETWORK_PASSPHRASE,
   base_fee=base_fee
)');
//   for($i=0;$i<count($trusts);$i++) {
   foreach ($trusts as $key => $tline) {
      $asset=explode(":",$tline);
      fwrite($gravar,'.append_change_trust_op(
   asset_code="' . $asset[0] . '", asset_issuer="' . $asset[1] . '"
)');
   }
   for ($i=0;$i<count($claims);$i++) {
      fwrite($gravar,'.append_claim_claimable_balance_op(
   balance_id="' . $claims[$i] . '",
   source="' . $account . '"
)');
   }
   fwrite($gravar,'
tx = transaction.build()');
   if ((substr($myPrivateKey,0,1) == "S") && (strlen($myPrivateKey) == 56)) {
      fwrite($gravar,'
stellar_keypair = Keypair.from_secret("' . $myPrivateKey . '")
account_priv_key = stellar_keypair.secret
tx.sign(account_priv_key)
response = server.submit_transaction(tx)
print(response["successful"], response["id"])');
   }
   else {
      fwrite($gravar,'
print(tx.to_xdr())');
   }
   system("python3 $caminho");
   unlink($caminho);
}

function secondsToTime($apartir_quando) {
  // From https://stackoverflow.com/a/19680778
   $ts_now=time();
   $seconds=$apartir_quando-$ts_now;
   $dtF = new \DateTime('@0');
   $dtT = new \DateTime("@$seconds");
   return $dtF->diff($dtT)->format('%ad, %hh %im %ss');
}


?>

File: auto_claim.py
import time
import locale
import requests
from datetime import datetime
import requests.packages.urllib3.exceptions
from stellar_sdk import Server, Keypair, MuxedAccount, TransactionBuilder, Network, TimeBounds, FeeBumpTransaction, ClaimPredicate, Claimant, Asset

requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)


headers_gerais = {
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.0.0 Safari/537.36',
    'accept-language': 'pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7',
    'accept': '*/*',
    'accept-encoding': 'gzip, deflate, br',
    'connection': 'keep-alive'
}

max_trustlines = 10


def number_format(num, places=0):
    return locale.format_string('%.*f', (places, num), True)

def check_cotacao(org_code, org_issuer, qtd, minimo, dest_code, dest_issuer=None):
    if (org_code == 'XLM'):
        org_type = 'native'

    elif len(org_code) <= 4:
        org_type = 'credit_alphanum4'
    else:
        org_type = 'credit_alphanum12'

    if dest_code == 'XLM':
        cotacao_url = f'https://horizon.stellar.org/paths/strict-send?destination_assets=native&source_asset_type={org_type}&source_asset_code={org_code}&source_asset_issuer={org_issuer}&source_amount={qtd}'
    else:
        cotacao_url = f'https://horizon.stellar.org/paths/strict-send?destination_assets={dest_code}%3A{dest_issuer}&source_asset_type={org_type}&source_asset_issuer={org_issuer}&source_asset_code={org_code}&source_amount={qtd}'

    get_cotacao = requests.get(
        url=cotacao_url,
        headers=headers_gerais,
        timeout=60,
        verify=False
    )

    cotacao = get_cotacao.json()

    # print(cotacao)
    
    if len(cotacao) > 0:
        # print(cotacao['_embedded']['records'][0]['destination_amount'])
        # print(number_format(0.0000100, 7), minimo)
        valor_mercado = cotacao['_embedded']['records'][0]['destination_amount']

        if float(valor_mercado) > minimo:
            valor_mercado = number_format(float(valor_mercado), 7)

            print(f' 💰{valor_mercado} {dest_code}', end='')

            return valor_mercado
        else:
            valor_mercado = number_format(float(valor_mercado), 7)

            print(f' 🧨{valor_mercado} {dest_code}')

            return 0
    else:
      return 0

def proceed_trans(account, trusts, claims, valores_claims, myPrivateKey=None):
    # print(account, trusts, claims, valores_claims, myPrivateKey)

    server = Server('https://horizon.stellar.org')
    print('Gerando XDR...')

    base_fee = 120

    muxed = MuxedAccount(account_id=account, account_muxed_id=1)
    print(muxed.account_muxed)
    muxed = MuxedAccount.from_account(muxed.account_muxed)
    
    account = server.load_account(account)

    print(f'account_id: {muxed.account_id}\naccount_muxed_id: {muxed.account_muxed_id}')

    native_asset = Asset('XLM')

    path = {}
    assets = {}

    for key, clx in valores_claims.items():
        if clx['dest_code'] == 'XLM' or clx['dest_code'] == '' or 'dest_code' not in clx:
            asset_to_sell = Asset(clx['org_code'], clx['org_issuer'])
            path_payments = Server.strict_send_paths(server, source_asset=asset_to_sell, source_amount=clx['org_valor'], destination=[native_asset]).call()
            path[key] = [Asset('XLM') for asset in path_payments['_embedded']['records']]
        else:
            asset_to_sell = Asset(clx['org_code'], clx['org_issuer'])
            path_payments = Server.strict_send_paths(server, source_asset=asset_to_sell, source_amount=clx['org_valor'], destination=[native_asset]).call()
            path[key] = [asset_to_buy for asset in path_payments['_embedded']['records']]

    transaction = TransactionBuilder(
       source_account=account,
       network_passphrase=Network.PUBLIC_NETWORK_PASSPHRASE,
       base_fee=base_fee
    )

    transaction.add_time_bounds(int(time.time()) - 60, int(time.time()) + 300)

    for key, tline in trusts.items():
        asset = tline.split(':')

        assets[asset[0]] = Asset(asset[0], asset[1])
        
        transaction.append_change_trust_op(asset=assets[asset[0]])

    for claim in claims:
        transaction.append_claim_claimable_balance_op(balance_id=claim, source=muxed)

    d = 0
    for key, clx in valores_claims.items():
        if clx['dest_code'] == 'XLM':
            clx['dest_code'] = 'XLM'
            clx['dest_issuer'] = None
            clx['dest_min'] = '0.0001'
        else:
            clx['dest_issuer'] = clx['dest_issuer']

        assets[clx['org_code']] = Asset(clx['org_code'], clx['org_issuer'])
        assets[clx['dest_code']] = Asset(clx['dest_code'], clx['dest_issuer'])

        transaction.append_path_payment_strict_send_op(
            destination=muxed, # account,
            send_asset=assets[clx['org_code']],
            send_amount=clx['org_valor'],
            dest_asset=assets[clx['dest_code']],
            dest_min=clx['dest_min'],
            path=path[key]
        )
        
        transaction.append_change_trust_op(
            asset=assets[clx['org_code']],
            limit='0'
        )

    tx = transaction.build()

    if myPrivateKey[0:1] == 'S' and len(myPrivateKey) == 56:
        stellar_keypair = Keypair.from_secret(myPrivateKey)
        account_priv_key = stellar_keypair.secret
        tx.sign(account_priv_key)
        response = server.submit_transaction(tx)
    
        print('Funcionou?', response['successful'], response['id'])
    else:
        print(tx.to_xdr())


def verificar_conta(public_address, private_address):
    if public_address[0:1] != 'G' or len(public_address) != 56:
        print('Não é um endereço de carteira.')

        exit()

    if private_address[0:1] != 'S' or len(private_address) != 56:
        print('Não é um endereço de carteira privada.')

        exit()
        
    wlt = '[' + public_address[-10:] + ']'
    arrecadacao = 0
    
    get_claimable = requests.get(
        url=f'https://horizon.stellar.org/claimable_balances/?limit=200&claimant={public_address}',
        headers=headers_gerais,
        timeout=60,
        verify=False
    )

    claimable_data = get_claimable.json()

    claimable_records = claimable_data['_embedded']['records']

    if len(claimable_records) < 1:
        print(f'\n{wlt} 0️⃣')
    else:
        trustlines = {}
        ids_claims = []
        valor_claims = {}

        get_account = requests.get(
            url=f'https://horizon.stellar.org/accounts/{public_address}',
            headers=headers_gerais,
            timeout=60,
            verify=False
        )

        account_data = get_account.json()

        account_balances = account_data['balances']

        print(f'\n{wlt}')

        for claimable in claimable_records:
            asset = claimable['asset'].split(':')
            print(' (' + asset[0] + ') ', end='')
            
            possui_trustline = False

            for balance in account_balances:
                if 'asset_code' in balance and balance['asset_code'] == asset[0] and 'asset_issuer' in balance and balance['asset_issuer'] == asset[1]:
                    print('☑', end='')
                    # print('Já tem trustline: ' + claimable["id"] + '\n');
                    possui_trustline = True

                    break

            pode_claim = False
            claimants = claimable['claimants']

            for claimant in claimants:
                if claimant['destination'] == public_address:
                    if 'unconditional' in claimant['predicate'] and claimant['predicate']['unconditional']:
                        print('☑ (incondicional)', end='')
                        # print('❎', end='')
                        pode_claim = True
                    elif 'abs_before' in claimant['predicate']:
                        ate_quando = datetime.fromisoformat(claimant['predicate']['abs_before'].replace('Z', ''))
                        
                        if ate_quando < datetime.now():
                            pode_claim = False
                            print('❎ (expirou)', end='')
                        else:
                            pode_claim = True
                            print('☑ (valido)', end='')

                    if pode_claim:
                        get_asset = requests.get(
                            url=f'https://horizon.stellar.org/assets?asset_code={asset[0]}&asset_issuer={asset[1]}',
                            headers=headers_gerais,
                            timeout=60,
                            verify=False
                        )

                        asset_data = get_asset.json()

                        asset_records = asset_data['_embedded']['records'][0]

                        if not asset_records['flags']['auth_required']:
                            pode_claim = True
                        else:
                            print('✋ (REQUER AUTORIZACAO!)')
                            pode_claim = False
            if not pode_claim:
                print(' (nao pode claim)')

            if not possui_trustline and pode_claim:
                if len(trustlines) < max_trustlines:
                    valor_mercado = check_cotacao(asset[0], asset[1], claimable['amount'], 0.0001, 'XLM', None)

                    if float(valor_mercado) > 0.0001:
                        print('☑ (valido)')
                        arrecadacao += float(valor_mercado)

                        # print('🐧' + claimable['id'])

                        id_asset = claimable['asset'].replace(':', '_')

                        trustlines[id_asset] = claimable['asset']

                        claim_id = claimable['id']

                        ids_claims.append(claim_id)

                        valor_claims[claim_id] = {
                            'org_valor': claimable['amount'],
                            'org_code': asset[0],
                            'org_issuer': asset[1],
                            'dest_code': 'XLM',
                            'valor_mercado': valor_mercado
                        }
                else:
                    print('🧪 (max trustlines')
                    break # TODO: ???????????

    if len(ids_claims) > 0:
        print(f'\n\033[1;33m{wlt} Arrecadacao: {number_format(arrecadacao, 7)}\033[0m\n')
        
        proceed_trans(public_address, trustlines, ids_claims, valor_claims, private_address)
    else:
        print(f'\n\033[1;33m{wlt} 0 Arrecadacao\033[0m\n')


contas = {
    'public_address': 'private_address'
}

if __name__ == '__main__':
    for public_address, private_address in contas.items():
        verificar_conta(public_address, private_address)
